Meetup 21 Resources (#178)
w3c/wot-cg: Meetups/21/2024-11-14-minutes.md

## Meetup

### Introduction

Ege: Welcome to the 21st meetup!

### News

Ege: Next event we will host a meetup about using Web of Things tech in Scientific Instrumentation
Ege: the week after we will have WoT week a physical event dedicated to WoT. 
Ege: let us know if you want to present something!

#### WoT Week

Ege: We will see each other from 25-29 November in Munich

Ege: During the event we will host a Plugfest event an interoperability hackathon!

Ege: Wednesday it is demo time, you can show your WoT solution.

Ege: let us know if you want to join!

### Web of Things for Every Thing

Ege: Today we have Christian who worked in a lot of open-source projects and with the Matrix protocol.

Chris: Web of Things is my hobby, what are the limits of IoT ? I started 6 years ago when I uploaded a thing using node.js using an express server.

Chris: mainly for experimentation

Chris: I started in the Mozilla community and now I'm moving to the W3C Web of Things.

Chris: Using node-wot is pretty convenient. It easy to connect a thing to the internet. You can support a lot of output format and protocols. Really good for production.

Chris: but what are the limit of IoT.

> Chris: It all started with a simple application capable of getting information from Open Street map and create a thing for every element there.
> 
> Chris: however, the whole Open Street map is quite heavy. Therefore, I opted to use a lazy loading approach.
> 
> Chris: I required a lot of effort and I abandoned the project. I still like to improve it and therefore
> I created a service called WoT for anything.
> 
> Chris: What a bike station looks like in WoT for anything?
> 
> > Christian showing possible paths for city bike parking

Reviewer: ```suggestion
Chris: It all started with a simple application capable of getting information from Open Street Map and creating a thing for every element there.

Chris: However, the whole Open Street Map is quite heavy. Therefore, I opted to use a lazy loading approach.

Chris: It required a lot of effort, and I abandoned the project. I still like to improve it, and therefore, I created a service called _WoT for anything_.

Chris: What does a bike station look like in _WoT for anything_?

> Christian showing possible paths for city bike parking
```

> Chris: I defined basic API operations that maps more or less WoT native operations. Everything is lazy loaded.
> 
> > Christian shows a demo
> 
> Chris: I tried to connect lights and digital drum kit.
> 
> Chris: in this setup is not easy to map this kind of devices but I built a path that I can use to load specific features. In the end even if the UI is better all the data is transformed and send over a single channel.

Reviewer: ```suggestion
Chris: I defined basic API operations that more or less map WoT native operations. Everything is lazy loaded.

> Christian shows a demo

Chris: I tried to connect lights and a digital drum kit.

Chris: In this setup, it is not easy to map these kinds of device, but I built a path that I can use to load specific features. In the end, even if the UI is better, all the data is transformed and sent over a single channel.
```

> Chris: how did you test all of this?

Reviewer: Also, this seems to be a question asked of, not by, Chris. Confirm the speaker?
```suggestion
Chris: How did you test all of this?
```

> Chris: I created WoT Wrench. A generic UI capable to interact with things. I also played around with base64 media. This means I can play video or audio or upload an image. For e-paper displays I also down sampled images.
> 
> Chris: I also created a tool to track Mood in a long distance relationship, this simple demo is combining multiple values from two things and combining them. It is all dynamic.

Reviewer: ```suggestion
Chris: I created WoT Wrench. A generic UI capable of interacting with things. I also played around with base64 media. This means I can play video or audio or upload an image. For e-paper displays, I also down-sampled images.

Chris: I also created a tool to track Mood in a long distance relationship. This simple demo is retrieving multiple values from two things and combining them. It is all dynamic.
```

Chris: I've also played a little bit by creating different games.

> Chris: The whole presentation that I'm showing can be controlled as a WoT Thing
> 
> Chris: I have a great number of features already implemented but my free time is running out.

Reviewer: ```suggestion
Chris: The whole presentation that I'm showing can be controlled as a WoT Thing.

Chris: I have a number of great features already implemented, but my free time is running out.
```

> Christian shows features

> Chris: I still have a longer list of features that I want to implement.
> 
> Chris: If you want to help me in this effort you can ping me and send your things! Or just reach me online or at WoT Week.
> 
> Chris: That was all!
> 
> Ege: thank you!

Reviewer: ```suggestion
Chris: I still have a long list of features that I want to implement.

Chris: If you want to help me in this effort, you can ping me and send your things! Or just reach me online or at WoT Week.

Chris: That was all!

Ege: Thank you!
```

### Q&A

> Ege: how did you bridge demo work? how did you control the demo?
> 
> Chris: I used a second screen where I controlled the web thing.
> 
> Ege: nice to see a lot of different stuff, I think your presentation demonstrate that we can use Web of Things to model services.
> 
> Henk: Impressive, a little bit intimidating. How did you generated all the TDs for this examples? Have you automated it?

Reviewer: ```suggestion
Ege: How did you bridge demo work? How did you control the demo?

Chris: I used a second screen where I controlled the web thing.

Ege: Nice to see a lot of different stuff. I think your presentation demonstrates that we can use the Web of Things to model services.

Henk: Impressive, even a little bit intimidating. How did you generate all the TDs for these examples? Have you automated the TD generation?
```

> Chris: all written by hand. But TDs are lazy loaded and can be generated with variables. They are not that complicated.
> 
> Ege: About the light demo with channels, you showed that you come up with better interface. How did you do that? Can we generalize this approach?
> 
> Chris: is based in the URL but it is not really generalizable.
> 
> Cris: can you use Thing Model for this purpose?

Reviewer: ```suggestion
Chris: All written by hand. But TDs are lazy loaded and can be generated with variables. They are not that complicated.

Ege: About the light demo with channels — you showed that you came up with a better interface. How did you do that? Can we generalize this approach?

Chris: It's based in the URL but it is not really generalizable.

Cris: Can you use Thing Model for this purpose?
```

> Chris: yes but I still have to deal with the mapping at the server side.
> 
> Ege: It is not same affordance that is always forwarding to the same channel. Like single request splitting in two other actions in the backend.
> 
> Chris: not really, but I have actions that does this (write all 1s or 0s).
> 
> Henk: You've done a lot of work with properties, but what about events? have you worked with them?

Reviewer: ```suggestion
Chris: Yes, but I still have to deal with the mapping at the server side.

Ege: It is not one affordance that is always forwarding to the same channel? Like a single request splitting into two other actions in the backend.

Chris: Not really, but I have actions that do this (write all `1`s or `0`s).

Henk: You've done a lot of work with properties, but what about events? Have you worked with them?
```

> Chris: I haven't worked a lot with them, but I worked more with actions.
> 
> Ege: we will put more guidelines for deciding when to used event or observable properties.
> 
> Ege: You started this year and doing in your free time. How easy was to work with WoT, any friction?
> 
> Chris: I don't know how much I've spent on this, but way over 100 hours. In terms of difficulty, handle single properties was easy but handling multiple properties was harder.
> 
> Chris: Multiple events and query all actions is not clear. For example, the payloads of meta operations and how they should be implemented.

Reviewer: ```suggestion
Chris: I haven't worked a lot with them. I worked more with actions.

Ege: We will put more guidelines for deciding when to use events or observable properties.

Ege: You started this year and did it in your free time. How easy was it to work with WoT — any friction?

Chris: I don't know how much I've spent on this, but way over 100 hours. In terms of difficulty, handling single properties was easy; handling multiple properties was harder.

Chris: Multiple events and query-all-actions is not clear. For example, the payloads of meta operations, and how they should be implemented.
```

Ege: Is everything open source?

> Chris: Still have to decide the future on this project, the code is not really ready for production and I have decide the licensing. Working hard to getting this in useful state for others.
> 
> Ege: Any tips for UI generation?
> 
> Chris: Point out which libraries there are. It was fun for me, since I am interesting. Objects and Arrays are tricky, especially if they have more than two layers.
> 
> Ege: Can you show Godot Engine part a bit more?
> 
> Chris: (shows the game) It has a manual fetch and observe buttons.

Reviewer: ```suggestion
Chris: Still have to decide the future of this project. The code is not really ready for production, and I have to decide the licensing. Working hard to get this into a useful state for others.

Ege: Any tips for UI generation?

Chris: Point out which libraries there are. It was fun for me, since I am interested. Objects and Arrays are tricky, especially if they have more than two layers.

Ege: Can you show Godot Engine part a bit more?

Chris: (shows the game) It has manual fetch and observe buttons.
```

> Ege: How did you first hear about it?
> 
> Chris: Through Mozilla IoT. Started contributing to the project and got interested in IoT automation in general. Also really like WebSockets.
> 
> Henk: SSE is fine if you use HTTP/2. It is unidirectional of course.
> 
> Chris: Yes I saw the connection limit. If you do it lazy, you can only do 6 properties. I went for ws after that. I could not find a solution for the read and observe race condition.
> 
> Henk: In a hub, I can use the same connection for all the things in the hub. WS is probably the future though.
> 
> Chris: WS falls short in security part.

Reviewer: ```suggestion
Ege: How did you first hear about it?

Chris: Through Mozilla IoT. Started contributing to the project and got interested in IoT automation in general. Also really like WebSockets.

Henk: SSE is fine if you use HTTP/2. It is unidirectional, of course.

Chris: Yes, I saw the connection limit. If you do it lazy, you can only do 6 properties. I went for WS after that. I could not find a solution for the read-and-observe race condition.

Henk: In a hub, I can use the same connection for all the things in the hub. WS is probably the future though.

Chris: WS falls short in security.
```

## :ballot_box_with_check: Resolutions

## :exclamation: Action Items

> ## :envelope_with_arrow: Feedbacks
> 
> - More examples of meta operations are needed
> - Guidance for clients when there is read and observe -> First do observe and then read to make sure to get the current value. It is protocol dependent ofc -> Would be good for http binding.

Reviewer: ```suggestion
## :envelope_with_arrow: Feedback

- More examples of meta operations are needed.
- Guidance for clients when there is read and observe -> First do observe and then read, to make sure to get the current value. It is protocol dependent of course -> Would be good for HTTP binding.
```
